### PR TITLE
[Refactor]: optimized the Xline startup process

### PR DESCRIPTION
**Pull request by themanforfree**
Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)
optimized the Xline startup process
* what changes does this pull request make?
optimized the Xline startup process
* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)
no

### CODE REVIEW

File: xline/src/xline_utils/args.rs [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        CurpConfigBuilder, EngineConfig, InitialClusterState, LevelConfig, LogConfig,
+        RotationConfig, ServerTimeout, StorageConfig, TraceConfig, XlineServerConfig,
+    },
+    parse_batch_bytes, parse_duration, parse_log_level, parse_members, parse_rotation, parse_state,
+    ConfigFileError,
+};
+/// Command line arguments
+#[derive(Parser, Debug)]
+#[clap(author, version, about, long_about = None)]
+pub struct ServerArgs {
```
Comment (bsbds) [inline xline/src/xline_utils/args.rs:26]: Should this go to the `utils` crate?

File: xline/src/xline_utils/args.rs [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+}
+
+/// Parse config from command line arguments or config file
+/// # Errors
+/// Return error if parse failed
+#[inline]
+pub async fn parse_config() -> Result<XlineServerConfig> {
+    if env::args_os().len() == 1 {
+        let path =
+            env::var("XLINE_SERVER_CONFIG").unwrap_or_else(|_| "/etc/xline_server.conf".to_owned());
```
Comment (bsbds) [inline xline/src/xline_utils/args.rs:267]: Use constant for both these strs?

File: xline-test-utils/src/lib.rs [hunk 3]
```diff
@@ -189,13 +176,14 @@ impl Cluster {
...
-        Some((encoding_key, decoding_key))
-    }
+    // TODO
+    // fn test_key_pair() -> Option<(EncodingKey, DecodingKey)> {
+    //     let private_key = include_bytes!("../private.pem");
+    //     let public_key = include_bytes!("../public.pem");
+    //     let encoding_key = EncodingKey::from_rsa_pem(private_key).unwrap();
+    //     let decoding_key = DecodingKey::from_rsa_pem(public_key).unwrap();
+    //     Some((encoding_key, decoding_key))
+    // }
```
Comment (bsbds) [inline xline-test-utils/src/lib.rs:186]: Please remove these comments.

File: xline/src/xline_utils/mod.rs [hunk 4]
```diff
@@ -0,0 +1,7 @@
+/// Xline command line arguments
+mod args;
+/// Xline tracing init
+mod init_trace;
```
Comment (iGxnon) [inline xline/src/xline_utils/mod.rs:4]: rename to `trace`

File: xline/tests/it/auth_test.rs [hunk 5]
```diff
@@ -104,7 +108,8 @@ async fn test_auth_non_authorized_rpcs() -> Result<(), Box<dyn Error>> {
 #[tokio::test(flavor = "multi_thread")]
 #[abort_on_panic]
 async fn test_kv_authorization() -> Result<(), Box<dyn Error>> {
-    let mut cluster = Cluster::new(3).await;
+    configs_with_auth(3);
```
Comment (iGxnon) [inline xline/tests/it/auth_test.rs:111]: remove this line

File: xline/src/server/xline_server.rs [hunk 6]
```diff
@@ -57,52 +72,88 @@ type CurpServer<S> = Rpc<Command, State<S, Arc<CurpClient>>>;
...
+        let name = cluster_config.name().clone();
+        let all_members = cluster_config.members().clone();
+        let init_cluster_info = ClusterInfo::new(all_members, &name);
+        Ok(match *cluster_config.initial_cluster_state() {
+            InitialClusterState::New => init_cluster_info,
+            InitialClusterState::Existing => get_cluster_info_from_remote(
+                &init_cluster_info,
+                server_addr_str,
+                &name,
+                Duration::from_secs(3),
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/server/xline_server.rs]: Fixed

[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/server/xline_server.rs]: Fixed

File: xline/src/server/xline_server.rs [hunk 7]
```diff
@@ -503,6 +587,40 @@ impl XlineServer {
     pub async fn stop(&self) {
         self.shutdown_trigger.self_shutdown_and_wait().await;
     }
+
+    /// Read key pair from file
+    async fn read_key_pair(
+        private_key_path: Option<&PathBuf>,
+        public_key_path: Option<&PathBuf>,
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/server/xline_server.rs:594]: Refactored

File: xline/src/server/xline_server.rs [hunk 8]
```diff
@@ -503,6 +587,40 @@ impl XlineServer {
...
+    ) -> Option<(EncodingKey, DecodingKey)> {
+        let encoding_key = match fs::read(private_key_path?).await {
+            Ok(key) => match EncodingKey::from_rsa_pem(&key) {
+                Ok(key) => key,
+                Err(e) => {
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/server/xline_server.rs]: Refactored

File: xline/src/xline_utils/args.rs
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/xline_utils/args.rs:1]: Renamed

File: xline/src/xline_utils/args.rs (hunk 1 above)
Comment (themanforfree) [reply xline/src/xline_utils/args.rs]: This struct only used by xline, it shouldn't move to utils

File: xline/src/xline_utils/args.rs (hunk 2 above)
Comment (themanforfree) [reply xline/src/xline_utils/args.rs]: Done

File: xline-test-utils/src/lib.rs (hunk 3 above)
Comment (themanforfree) [reply xline-test-utils/src/lib.rs]: Removed

File: xline/src/xline_utils/mod.rs (hunk 4 above)
Comment (themanforfree) [reply xline/src/xline_utils/mod.rs]: Renamed

File: xline/tests/it/auth_test.rs (hunk 5 above)
Comment (themanforfree) [reply xline/tests/it/auth_test.rs]: Removed

File: xline/src/server/xline_server.rs [hunk 9]
```diff
@@ -57,52 +72,88 @@ type CurpServer<S> = Rpc<Command, State<S, Arc<CurpClient>>>;
...
+                };
+                address.to_socket_addrs()
+            })
+            .collect::<Result<Vec<_>, _>>()?
+            .into_iter()
+            .flatten()
+            .collect_vec();
+        debug!("name = {:?}", cluster_config.name());
+        debug!("server_addr = {server_addr:?}");
+        debug!("cluster_peers = {:?}", cluster_config.members());
```
Comment (Phoenix500526) [inline xline/src/server/xline_server.rs:140]: It's better to use `info!` here.

File: xline/src/utils/trace.rs [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+use tracing_subscriber::fmt::format;
+use tracing_subscriber::util::SubscriberInitExt;
+use tracing_subscriber::{layer::SubscriberExt, Layer};
+use utils::config::{file_appender, LogConfig, TraceConfig};
+
+/// init tracing subscriber
+/// # Errors
+/// Return error if init failed
+#[inline]
+pub fn init_subscriber(
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/utils/trace.rs:14]: Removed

File: xline/src/server/xline_server.rs (hunk 9 above)
Comment (themanforfree) [reply xline/src/server/xline_server.rs]: Modified

**Review by Phoenix500526 — APPROVED** [review]
LGTM

**Review by bsbds — APPROVED** [review]

**Pull request merged**